nservicebus/nhibernate/custom-mappings.md

```diff
@@ -0,0 +1,125 @@
+ - nservicebus/nhibernate/accessing-data
+ - nservicebus/nhibernate/index
+---
+
+
+Sometimes the database schema that gets generated by `NServiceBus.NHibernate` for your saga data types is insufficient. For example, you might need to store strings with a different string length, want so use complex types, need to store DateTime values with a high precision or maybe want to tweak the eager versus lazy loading to improve performance. You will need to create a custom mapping to achieve this so that the generated default mapping will not be used.
+
+Custom mapping options:
+
+* Create a `*.hbm.xml` file for your saga data class
+* Use Fluent NHibernate (popular separate fluent api)
+* Use NHibernate.Mapping.Attributes
+* Loquacious Configuration (native fluent api)
+
+There are probably even other options but these are the most frequently used.
+
+See [NHibernate configuration](configuration) on how you can pass this configuration object to the NServiceBus NHibernate configuration.
+
+NOTE: When you create a custom mapping then you are responsible for mapping the primary key and unique indexes correctly.
+
+
+NOTE: The `[Unique]` attribute will not be used to generate a unique index. You are responsible to add this to your custom mapping. This is especially important when using both optimistic concurrency and a transaction isolation level different from serializable and allow concurrent processing of messages. Not adding this can result in multiple saga entities be persisted as the second insert will not fail when inserting a duplicate saga key (different from the primary key column) when multiple message  are processed concurrently that target the same saga instance.
+
+
+## Custom .hbm.xml mapping
+
+Using NHibernate mapping files is the original way to customize your mapping. It needs to be created and then either embedded as a resource in your assembly or available on the file system.
+
+Look at these pages to see how you can create these mappings:
+
+* http://nhibernate.info/doc/tutorials/first-nh-app/your-first-nhibernate-based-application.html
+
+
+### Mappings as embedded resources
+
+It is not required to create your own NHibernate configuration object and pass it to the NServiceBus NHibernate configuration.
+
+
+### Read mappings from file system instead of embedded resources
+
+If you do not want to embed the `*.hbm.xml` files then you can also load mappings from the file system. Create a custom NHibernate configuration object and use the following example to add mappings from the file system.
+
+<!-- import NHibernateLoadMappingsFromFileSystem -->
+
+## Use Fluent NHibernate
+
+Fluent NHibernate gives you a type-safe mapping approach where the mapping is specified as c# code and not as xml but still seperate from the classes. The benefit is that you get compile time feedback when a mapping is not valid any more when you have breaking changes in your mapping but also that your classes and mappings are not part of the same .net type.
```

Comment:
I'd say no, we should not include any detail from the NH side of things unless strictly necessary, otherwise we end up maintaining NH documentation on not ours.